spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputePartitionStatsProcedure.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.spark.sql.connector.iceberg.catalog.ProcedureParameter;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A procedure that computes the stats incrementally after the snapshot that has partition stats
+ * file till the given snapshot (uses current snapshot if not specified) and writes the combined
+ * result into a {@link PartitionStatisticsFile} after merging the stats for a given snapshot. Does
+ * a full compute if previous statistics file does not exist. Also registers the {@link
+ * PartitionStatisticsFile} to table metadata.
+ *
+ * @see SparkActions#computePartitionStats(Table)
+ */
+public class ComputePartitionStatsProcedure extends BaseProcedure {
+
+  private static final ProcedureParameter TABLE_PARAM =
+      ProcedureParameter.required("table", DataTypes.StringType);
+  private static final ProcedureParameter SNAPSHOT_ID_PARAM =
+      ProcedureParameter.optional("snapshot_id", DataTypes.LongType);
+  private static final ProcedureParameter FULL_REFRESH_PARAM =
+      ProcedureParameter.optional("full_refresh", DataTypes.BooleanType);
+
+  private static final ProcedureParameter[] PARAMETERS =
+      new ProcedureParameter[] {TABLE_PARAM, SNAPSHOT_ID_PARAM, FULL_REFRESH_PARAM};
+
+  private static final StructType OUTPUT_TYPE =
+      new StructType(
+          new StructField[] {
+            new StructField(
+                "partition_statistics_file", DataTypes.StringType, true, Metadata.empty())
+          });
+
+  public static ProcedureBuilder builder() {
+    return new Builder<ComputePartitionStatsProcedure>() {
+      @Override
+      protected ComputePartitionStatsProcedure doBuild() {
+        return new ComputePartitionStatsProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private ComputePartitionStatsProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    ProcedureInput input = new ProcedureInput(spark(), tableCatalog(), PARAMETERS, args);
+    Identifier tableIdent = input.ident(TABLE_PARAM);
+    Long snapshotId = input.asLong(SNAPSHOT_ID_PARAM, null);
+    Boolean fullRefresh = input.asBoolean(FULL_REFRESH_PARAM, false);
+
+    return modifyIcebergTable(
+        tableIdent,
+        table -> {
+          if (fullRefresh) {
+            UpdatePartitionStatistics updateStats = table.updatePartitionStatistics();
+            table
+                .partitionStatisticsFiles()
+                .forEach(file -> updateStats.removePartitionStatistics(file.snapshotId()));
+            updateStats.commit();
```

Comment:
Is this two separate trasnactions?  There's no way to have it all or nothing?  (to prevent the statistic from being removed if the compute fail)

Reply:
Yes. When we designed the original core API, we didn't provide an option to full refresh as it is very rare case for user to use it (only during corruption), that too when incremental compute can't read the previous stats file it fallback to full compute.  So, if at all for other reasons users need a full refresh, they need to unregister and call compute again. 

If two separate transaction is a bad idea. I can remove this option and we can add it in the future if required (if any new use case comes up). WDYT?

Reply:
Agree. I just removed the full refresh option now. 

Since the underlaying action will fallback to full compute if stats are corrupted automatically, we don't have use case for full refresh yet. If there are new use case in the future, we can support it from underlaying API to make it atomic instead of using remove stats and compute method. 

PR is ready now.